pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageContainerImpl.java

```diff
@@ -69,8 +69,7 @@ public boolean add(MessageImpl<?> msg, SendCallback callback) {
                 messageMetadata.setSequenceId(msg.getSequenceId());
                 lowestSequenceId = Commands.initBatchMessageMetadata(messageMetadata, msg.getMessageBuilder());
                 this.firstCallback = callback;
-                batchedMessageMetadataAndPayload = PulsarByteBufAllocator.DEFAULT
-                        .buffer(Math.min(maxBatchSize, ClientCnx.getMaxMessageSize()));
+                batchedMessageMetadataAndPayload = PulsarByteBufAllocator.DEFAULT.compositeBuffer();
```

Comment:
This approach could result in many buffer expansions when the average buffer size is larger than the "default" size of allocated buffers.

Reply:
Buffer expansions will happens in some situations, but for `CompositeByteBuf`, it wouldn't takes too much costs.
`mem_copy` operations triggered when `CompositeByteBuf#consolidate0`, which means after write more than 1000000 bytes into buffer, `mem_copy` will be triggered first time by default. I believe it's acceptable in the `client side`

Reply:
Could you please do your final comment @merlimat